wagtail/admin/api/endpoints.py

```diff
@@ -56,16 +47,20 @@ class PagesAdminAPIViewSet(PagesAPIViewSet):
         'has_children'
     ])
 
-    def get_queryset(self):
-        request = self.request
+    def get_root_page(self):
+        """
+        Returns the page that is used when the `&child_of=root` filter is used.
+        """
+        return Page.get_first_root_node()
 
-        # Allow pages to be filtered to a specific type
-        try:
-            models = page_models_from_string(request.GET.get('type', 'wagtailcore.Page'))
-        except (LookupError, ValueError):
-            raise BadRequestError("type doesn't exist")
+    def get_base_queryset(self, models=None):
```

Comment:
The `models` parameter is bothering me a bit here... I can see why it's a necessary evil (we need to be able to swap out the base model of the queryset so that we can filter on custom `APIField`s defined in the specific page model), but the "if len(models) == 1 then replace the base model, otherwise filter" logic in here feels like too much undocumented magic for something claiming to be a base queryset.

At the very least, I think this parameter needs to be mentioned in the docstring, because as it stands it's not obvious that the "secondary" uses of get_base_queryset (i.e. retrieving the root page for `child_of` / `descendant_of`, and calculating child/descendant counts in c67042922aa74dede6d26ab09baf6a8e48d63764) rely on us _not_ passing a specific page model here.

Beyond that, I think we could improve things a bit by having `get_base_queryset` accept a single model (defaulting to Page), and have the `if len(models)` logic in `get_queryset` so that `get_base_queryset` isn't doing any filtering (other than the implicit filtering that happens 'for free' by querying on a Page subclass, and that should hopefully be fairly clear from the method signature).

As a more drastic change (and I'm not sure how feasible this is) - perhaps instead of a `get_base_queryset`, we could have a `get_base_query_filters` method that returns a `Q()` object that the caller can apply to whatever queryset it wants?

Reply:
Done as per we discussed offline